apps/balancer-tools/src/app/apr/(utils)/tokenApr.ts

```diff
@@ -285,3 +321,159 @@ const getImplementationAddress = async (
     return undefined;
   }
 };
+
+async function calculateTokenApr({
+  tokenYield,
+  poolId,
+  chain,
+  tokenAddress,
+  tokenWeight,
+  timeStart,
+  timeEnd,
+}: {
+  tokenYield: number;
+  poolId: string;
+  chain: string;
+  tokenAddress: Address;
+  tokenWeight: number;
+  timeStart: number;
+  timeEnd: number;
+}) {
+  const rangeData = await poolsWithCache.gql(chain).poolSnapshotInRange({
+    poolId,
+    timestamp: generateDateRange(timeStart, timeEnd),
+  });
+
+  let pool: Pool;
+
+  if (rangeData.poolSnapshots.length > 0) {
+    pool = rangeData.poolSnapshots[0].pool;
+  } else {
+    const currentPoolData = await pools.gql(chain).Pool({ poolId });
+    if (!currentPoolData || !currentPoolData.pool) {
+      throw new Error("Pool data is not found or is invalid.");
+    }
+    pool = currentPoolData.pool;
+  }
+
+  if (!pool || !pool.tokens) {
+    return 0;
+  }
```

Comment:
Not sure what's the use case that makes us need this here (and I ask myself if we would've needed it elsewhere) but you'd help your reader if you encapsulate this in another function:

```
async function fetchPoolData(chain: string, poolId: string, timeStart: number, timeEnd: number): Promise<Pool | null> {
  const rangeData = await poolsWithCache.gql(chain).poolSnapshotInRange({
    poolId,
    timestamp: generateDateRange(timeStart, timeEnd),
  });
  if (rangeData.poolSnapshots.length > 0) {
    return rangeData.poolSnapshots[0].pool;
  }
  const currentPoolData = await pools.gql(chain).Pool({ poolId });
  return currentPoolData?.pool || null;
}

// ... then in calculateTokenApr

...
  const pool = await fetchPoolData(chain, poolId, timeStart, timeEnd);
  if (!pool || !pool.tokens) return 0;
...
```